pkg/descheduler/strategies/topologyspreadconstraint.go

```diff
@@ -0,0 +1,220 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package strategies
+
+import (
+	"context"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	"math"
+	"sigs.k8s.io/descheduler/pkg/api"
+	"sigs.k8s.io/descheduler/pkg/descheduler/evictions"
+	"sort"
+)
+
+// AntiAffinityTerm's topology key value used in predicate metadata
+type topologyPair struct {
+	key   string
+	value string
+}
+
+type topology struct {
+	pair topologyPair
+	pods []*v1.Pod
+}
+
+func RemovePodsViolatingTopologySpreadConstraint(
+	ctx context.Context,
+	client clientset.Interface,
+	strategy api.DeschedulerStrategy,
+	nodes []*v1.Node,
+	podEvictor *evictions.PodEvictor,
+) {
+	nodeMap := make(map[string]*v1.Node)
+	for _, node := range nodes {
+		nodeMap[node.Name] = node
+	}
+
+	// 1. for each namespace for which there is Topology Constraint
+	// 2. for each TopologySpreadyConstraint in that namespace
+	//  { find all evictable pods in that namespace
+	//  { 3. for each evictable pod in that namespace
+	// 4. If the pod matches this TopologySpreadConstraint LabelSelector
+	// 5. If the pod nodeName is present in the nodeMap
+	// 6. create a topoPair with key as this TopologySpreadConstraint.TopologyKey and value as this pod's Node Label Value for this TopologyKey
+	// 7. add the pod with key as this topoPair
+	// 8. find the min number of pods in any topoPair for this topologyKey
+	// iterate through all topoPairs for this topologyKey and diff currentPods -minPods <=maxSkew
+	// if diff > maxSkew, add this pod in the current bucket for eviction
+
+	evictable := podEvictor.Evictable()
+
+	namespacedTopologySpreadConstrainPods := make(map[string][]v1.Pod)
+	// First record all of the constraints by namespace
+	namespaces, err := client.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		klog.ErrorS(err, "Couldn't list namespaces")
+		return
+	}
+	podsForEviction := make(map[*v1.Pod]struct{})
+	// 1. for each namespace...
+	for _, namespace := range namespaces.Items {
+		namespacePods, err := client.CoreV1().Pods(namespace.Name).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			klog.ErrorS(err, "Couldn't list pods in namespace", "namespace", namespace)
+			return
+		}
+
+		// ...where there is a topology constraint
+		for _, pod := range namespacePods.Items {
+			if pod.Spec.TopologySpreadConstraints != nil {
+				namespacedTopologySpreadConstrainPods[pod.Namespace] = append(namespacedTopologySpreadConstrainPods[pod.Namespace], pod)
+			}
+		}
+		if len(namespacedTopologySpreadConstrainPods[namespace.Name]) == 0 {
+			continue
+		}
+
+		// 2. for each topologySpreadConstraint in that namespace
```

Comment:
Yeah that's a good question. I kind of intentionally just went with checking all constraints here because I wasn't sure how we should handle hard vs soft. Hard is obviously a requirement but maybe a strategy parameter to consider soft constraints too? (ie `softConstraints: true`)

Then, should the 2 be handled any differently? It seems like hard constraints should be evaluated first but I don't think it makes a difference

Reply:
I'd suggest starting with hard constraints. The semantics of soft constraints is a bit complex, you have to strictly follow up with the upstream code to make it compliant, which I think is a good-to-have for descheduelr, at least for now.